File: ExampleAnalysis.m
close all;
clear all;

%this file runs plots the results of a click simulation performed with
%ExampleSimulation.m
load Simulations.mat

%are the different conditions
L=[0 10 20 30 40 50 60 70 80 90 100];
p0=2e-5;
CF=output.cf;
figure,plot(CF),xlabel('Cochlear Channel Number [-]'),ylabel('Characteristic Frequency [Hz]')

fs_c=output.fs_bm; %the sampling frequency of BM, OAE and IHC are higher to avoid numerical errors (see Altoe et al., 2014 JASA)
fs=output.fs_an; %the sampling frequency of AN, CN, IC and waves are 5 times lower
t_c=[0:size(output(1).v,1)-1]/fs_c;
t=[0:size(output(1).anfH,1)-1]/fs;
f_c=0:fs_c/size(output(1).v,1):fs_c-1/size(output(1).v,1);
f=0:fs/size(output(1).anfH,1):fs-1/size(output(1).anfH,1);

% pick a channel number to plot results from. The CF corresponding to the
% channel number depends on whether you chose 'all', 'half' or 'abr'
No=245;

%reorganisation of the data for easier processing
for n=1:numel(L)
   oae(:,n)=output(n).e;
   vrms(:,n)=rms(output(n).v);
   ihcrms(:,n)=rms(output(n).ihc);
   v(:,n)=output(n).v(:,No);
   ihc(:,n)=output(n).ihc(:,No);
end

%% the OAE
%this figure is the ear-canal pressure: 
%to get the reflection-source OAE, do the following simulation:
%OAE_{reflections on}-OAE_{reflections off}
%to get the distortion-source OAE, do a normalisation using a low level
%(linear) simulation that the reflections off
figure,
subplot(2,1,1),plot(1000*t_c,oae)
xlabel('Time [ms]'),ylabel('Ear Canal Pressure [Pa]'),xlim([0 20]),ylim([-0.02 0.02]),legend(num2str(L')),legend('boxoff')
oae(1:200,:)=0; %zeropadding to plot OAE spectrum
subplot(2,1,2),plot(f_c/1000,20*log10(abs(fft(oae/p0))));
xlabel('Frequency [kHz]'),ylabel('EC Magnitude [dB re p0]'),xlim([0 12]),legend(num2str(L')),legend('boxoff')

%% v_bm and V_IHC 
figure,
subplot(2,2,1),plot(1000*t_c,v),
xlabel('Time [ms]'),ylabel('v_{bm} [m/s]'),xlim([0 30]),legend(num2str(L')),legend('boxoff')
title(['CF of ',num2str(round(CF(No))),' Hz'])
subplot(2,2,2),plot(CF/1000,20*log10(vrms)),
xlabel('CF [kHz]'),ylabel('rms of v_{bm} [dB re 1 m/s]'),xlim([0 14]),ylim([max(max(20*log10(vrms)))-100 max(max(20*log10(vrms)))+10])
title(['Excitation Pattern'])
subplot(2,2,3),plot(1000*t_c,ihc),
xlabel('Time [ms]'),ylabel('V_{ihc} [V]'),xlim([0 30]),legend(num2str(L')),legend('boxoff')
title(['CF of ',num2str(round(CF(No))),' Hz'])
subplot(2,2,4),plot(CF/1000,ihcrms),
xlabel('CF [kHz]'),ylabel('rms of V_{ihc} [V]'),xlim([0 14]),%ylim([max(max(20*log10(ihcrms)))-100 max(max(20*log10(ihcrms)))+10])
title(['Excitation Pattern'])

%reorganisation of the data for easier processing
for n=1:numel(L)
   HSR(:,n)=output(n).anfH(:,No);
   MSR(:,n)=output(n).anfM(:,No);
   LSR(:,n)=output(n).anfL(:,No);
   AN(:,n)=output(n).an_summed(:,No);
   CN(:,n)=output(n).cn(:,No);
   IC(:,n)=output(n).ic(:,No);
   W1(:,n)=output(n).w1;
   W3(:,n)=output(n).w3;
   W5(:,n)=output(n).w5;
   EFR(:,n)=output(n).w1+output(n).w3+output(n).w5;
end

%single unit responses
figure,
subplot(3,2,1),plot(1000*t,HSR)
title(['CF of ',num2str(round(CF(No))),' Hz'])
xlim([0 20]),xlabel('Time [ms]'),ylabel('HSR fiber [spikes/s]')
legend(num2str(L')),legend('boxoff')
subplot(3,2,3),plot(1000*t,MSR)
xlim([0 20]),xlabel('Time [ms]'),ylabel('MSR fiber [spikes/s]')
subplot(3,2,5),plot(1000*t,LSR)
xlim([0 20]),xlabel('Time [ms]'),ylabel('LSR fiber [spikes/s]')

subplot(3,2,2),plot(1000*t,AN)
title(['CF of ',num2str(round(CF(No))),' Hz'])
xlim([0 20]),xlabel('Time [ms]'),ylabel('sum AN [spikes/s]')
%Spikes summed across all fibers @ 1 CF
subplot(3,2,4),plot(1000*t,CN)
xlim([0 20]),xlabel('Time [ms]'),ylabel('CN [spikes/s]')
subplot(3,2,6),plot(1000*t,IC)
xlim([0 20]),xlabel('Time [ms]'),ylabel('IC [spikes/s]')

%population responses
figure
subplot(4,1,1),plot(1000*t,1e6*W1)
title(['Population Responses summed across simulated CFs'])
xlim([0 20]),xlabel('Time [ms]'),ylabel('W-1 [\muV]')
legend(num2str(L')),legend('boxoff')
subplot(4,1,2),plot(1000*t,1e6*W3)
xlim([0 20]),xlabel('Time [ms]'),ylabel('W-3 [\muV]')
subplot(4,1,3),plot(1000*t,1e6*W5)
xlim([0 20]),xlabel('Time [ms]'),ylabel('W-5 [\muV]')
subplot(4,1,4),plot(1000*t,1e6*EFR)
xlim([0 20]),xlabel('Time [ms]'),ylabel('EFR [\muV]')

%The model code and interface was written by Alessandro Altoè and Sarah Verhulst (copyright 2012,2014,2015,2016,2018) 
%and is licensed under the UGent acadamic license (see details in license file that is part of this repository). 
%The Verhulstetal2018Model consists of the following files: 
%tridiag.so, cochlea_utils.c, run_model2018.py, model2018.m, cochlear_model2017.py, inner_hair_cell2018.py, auditory_nerve2017.py, ic_cn2017.py, ExampleSimulation.m, ExampleAnalysis.m, the HI profiles in the Poles folder. 
 

File: ExampleAnalysis.py
import numpy as np
import matplotlib.pyplot as plt
import scipy.io as sio
from scipy import signal
import os

# This file runs and plots the results of a click simulation performed with
# ExampleSimulation.py
# Load the simulation results
data = sio.loadmat('Simulations.mat')
output = data['output'][0]  # MATLAB struct becomes a structured array

# Define the different conditions
L = [0]
p0 = 2e-5
CF = output['cf'].item().flatten()  # Handle nested array structure

# Create figure for CF plot
plt.figure()
plt.plot(CF)
plt.xlabel('Cochlear Channel Number [-]')
plt.ylabel('Characteristic Frequency [Hz]')
plt.title('Characteristic Frequencies')
plt.grid(True)
plt.show()

# Sampling frequencies
fs_c = output['fs_bm'].item().item()  # BM, OAE and IHC sampling frequency (higher to avoid numerical errors)
fs = output['fs_an'].item().item()    # AN, CN, IC and waves sampling frequency (5 times lower)

# Time vectors
t_c = np.arange(output['v'].item().shape[0]) / fs_c
t = np.arange(output['anfH'].item().shape[0]) / fs

# Frequency vectors
f_c = np.arange(output['v'].item().shape[0]) * fs_c / output['v'].item().shape[0]
f = np.arange(output['anfH'].item().shape[0]) * fs / output['anfH'].item().shape[0]

# Pick a channel number to plot results from
No = 245

# Reorganization of the data for easier processing
num_conditions = len(L)
v_data = output['v'].item()
ihc_data = output['ihc'].item()
e_data = output['e'].item()

oae = np.zeros((e_data.shape[1], num_conditions))
vrms = np.zeros((v_data.shape[1], num_conditions))
ihcrms = np.zeros((ihc_data.shape[1], num_conditions))
v = np.zeros((v_data.shape[0], num_conditions))
ihc = np.zeros((ihc_data.shape[0], num_conditions))

for n in range(num_conditions):
    oae[:, n] = e_data.flatten()
    vrms[:, n] = np.sqrt(np.mean(v_data**2, axis=0))
    ihcrms[:, n] = np.sqrt(np.mean(ihc_data**2, axis=0))
    v[:, n] = v_data[:, No]
    ihc[:, n] = ihc_data[:, No]

# The OAE (Otoacoustic Emission)
# This figure shows the ear-canal pressure:
# To get the reflection-source OAE, do the following simulation:
# OAE_{reflections on} - OAE_{reflections off}
# To get the distortion-source OAE, do a normalization using a low level
# (linear) simulation with reflections off
plt.figure(figsize=(10, 8))

plt.subplot(2, 1, 1)
plt.plot(1000 * t_c, oae)
plt.xlabel('Time [ms]')
plt.ylabel('Ear Canal Pressure [Pa]')
plt.xlim([0, 20])
plt.ylim([-0.02, 0.02])
plt.legend([str(l) for l in L], frameon=False)
plt.title('OAE Time Domain')

# Zero padding to plot OAE spectrum
oae_spectrum = oae.copy()
oae_spectrum[:200, :] = 0

plt.subplot(2, 1, 2)
oae_fft = np.fft.fft(oae_spectrum / p0, axis=0)
plt.plot(f_c / 1000, 20 * np.log10(np.abs(oae_fft)))
plt.xlabel('Frequency [kHz]')
plt.ylabel('EC Magnitude [dB re p0]')
plt.xlim([0, 12])
plt.legend([str(l) for l in L], frameon=False)
plt.title('OAE Frequency Domain')
plt.grid(True)
plt.tight_layout()
plt.show()

# v_bm and V_IHC
plt.figure(figsize=(12, 10))

plt.subplot(2, 2, 1)
plt.plot(1000 * t_c, v)
plt.xlabel('Time [ms]')
plt.ylabel('v_{bm} [m/s]')
plt.xlim([0, 30])
plt.legend([str(l) for l in L], frameon=False)
plt.title(f'CF of {round(CF[No])} Hz')
plt.grid(True)

plt.subplot(2, 2, 2)
plt.plot(CF / 1000, 20 * np.log10(vrms))
plt.xlabel('CF [kHz]')
plt.ylabel('rms of v_{bm} [dB re 1 m/s]')
plt.xlim([0, 14])
plt.ylim([np.max(20 * np.log10(vrms)) - 100, np.max(20 * np.log10(vrms)) + 10])
plt.title('Excitation Pattern')
plt.grid(True)

plt.subplot(2, 2, 3)
plt.plot(1000 * t_c, ihc)
plt.xlabel('Time [ms]')
plt.ylabel('V_{ihc} [V]')
plt.xlim([0, 30])
plt.legend([str(l) for l in L], frameon=False)
plt.title(f'CF of {round(CF[No])} Hz')
plt.grid(True)

plt.subplot(2, 2, 4)
plt.plot(CF / 1000, ihcrms)
plt.xlabel('CF [kHz]')
plt.ylabel('rms of V_{ihc} [V]')
plt.xlim([0, 14])
plt.title('Excitation Pattern')
plt.grid(True)

plt.tight_layout()
plt.show()

# Reorganization of the data for easier processing
anfH_data = output['anfH'].item()
anfM_data = output['anfM'].item()
anfL_data = output['anfL'].item()
an_summed_data = output['an_summed'].item()
cn_data = output['cn'].item()
ic_data = output['ic'].item()
w1_data = output['w1'].item()
w3_data = output['w3'].item()
w5_data = output['w5'].item()

HSR = np.zeros((anfH_data.shape[0], num_conditions))
MSR = np.zeros((anfM_data.shape[0], num_conditions))
LSR = np.zeros((anfL_data.shape[0], num_conditions))
AN = np.zeros((an_summed_data.shape[0], num_conditions))
CN = np.zeros((cn_data.shape[0], num_conditions))
IC = np.zeros((ic_data.shape[0], num_conditions))
W1 = np.zeros((w1_data.shape[1], num_conditions))  # Use shape[1] for wave data
W3 = np.zeros((w3_data.shape[1], num_conditions))
W5 = np.zeros((w5_data.shape[1], num_conditions))
EFR = np.zeros((w1_data.shape[1], num_conditions))

for n in range(num_conditions):
    HSR[:, n] = anfH_data[:, No]
    MSR[:, n] = anfM_data[:, No]
    LSR[:, n] = anfL_data[:, No]
    AN[:, n] = an_summed_data[:, No]
    CN[:, n] = cn_data[:, No]
    IC[:, n] = ic_data[:, No]
    W1[:, n] = w1_data.flatten()
    W3[:, n] = w3_data.flatten()
    W5[:, n] = w5_data.flatten()
    EFR[:, n] = w1_data.flatten() + w3_data.flatten() + w5_data.flatten()

# Single unit responses
plt.figure(figsize=(12, 10))

plt.subplot(3, 2, 1)
plt.plot(1000 * t, HSR)
plt.title(f'CF of {round(CF[No])} Hz')
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('HSR fiber [spikes/s]')
plt.legend([str(l) for l in L], frameon=False)
plt.grid(True)

plt.subplot(3, 2, 3)
plt.plot(1000 * t, MSR)
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('MSR fiber [spikes/s]')
plt.grid(True)

plt.subplot(3, 2, 5)
plt.plot(1000 * t, LSR)
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('LSR fiber [spikes/s]')
plt.grid(True)

plt.subplot(3, 2, 2)
plt.plot(1000 * t, AN)
plt.title(f'CF of {round(CF[No])} Hz')
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('sum AN [spikes/s]')
plt.grid(True)

plt.subplot(3, 2, 4)
plt.plot(1000 * t, CN)
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('CN [spikes/s]')
plt.grid(True)

plt.subplot(3, 2, 6)
plt.plot(1000 * t, IC)
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('IC [spikes/s]')
plt.grid(True)

plt.tight_layout()
plt.show()

# Population responses
plt.figure(figsize=(10, 12))

plt.subplot(4, 1, 1)
plt.plot(1000 * t, 1e6 * W1)
plt.title('Population Responses summed across simulated CFs')
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('W-1 [μV]')
plt.legend([str(l) for l in L], frameon=False)
plt.grid(True)

plt.subplot(4, 1, 2)
plt.plot(1000 * t, 1e6 * W3)
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('W-3 [μV]')
plt.grid(True)

plt.subplot(4, 1, 3)
plt.plot(1000 * t, 1e6 * W5)
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('W-5 [μV]')
plt.grid(True)

plt.subplot(4, 1, 4)
plt.plot(1000 * t, 1e6 * EFR)
plt.xlim([0, 20])
plt.xlabel('Time [ms]')
plt.ylabel('EFR [μV]')
plt.grid(True)

plt.tight_layout()
plt.show()

print("Analysis complete!")